### PR TITLE
Add 'status' middleware instead of 'status' directive for 'rewrite' middleware

### PR DESCRIPTION
**Pull request by evvvvr**
Simple implementation of middleware for responding with specified status as discussed in #1012

### CODE REVIEW

**Comment by mholt** [comment]
Cool - thanks @evvvvr! Going to be looking through this soon.

**Comment by mholt** [comment]
This is looking nice so far, thank you @evvvvr. I would like @abiosoft to review it when he has a chance (I know he's busy).

File: caddyhttp/status/setup.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+		if _, exists := res[path]; exists {
+			return c.Errf("Duplicate path: '%s'", path)
+		}
+		res[path] = statusCode
+
+		return nil
+	}
+
+	for c.Next() {
+		hadBlock = false
```
Comment (mholt) [inline caddyhttp/status/setup.go]: Not sure we need block support do we? `status path code` is a pretty simple syntax. Do you think there will be lists of many items?

Comment (mholt) [inline caddyhttp/status/setup.go]: Okay, I just saw the tests below. I'm not sure the blocks are really worth the extra syntax complexity. _Maybe_ if they all shared the same status code... what do you think?

Comment (evvvvr) [inline caddyhttp/status/setup.go]: I made it by analogy with `errors` directive, where we can specify pages for different status codes in one block:
```
errors {
   what where
   what where... 
}
```  
IMO it is convenient to be able  to specify rules in one block, instead of writing the same directive multiple times like:
```
status path code
status path code...
```

Comment (abiosoft) [inline caddyhttp/status/setup.go]: Both can be supported.

Comment (mholt) [inline caddyhttp/status/setup.go]: Okay, so if you do allow blocks, what if it was for the same status code:

```
status 404 {
    /path1
    /path2
}
```

I figure there won't be so many different statuses that you'd have to write the word "status" a lot.

File: caddyhttp/status/setup.go [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+
+// statusParse parses status directive
+func statusParse(c *caddy.Controller) (map[string]int, error) {
+	res := make(map[string]int)
+	hadBlock := false
+
+	saveStatusRule := func(path, statusCodeString string) error {
+		statusCode, err := strconv.Atoi(statusCodeString)
+		if err != nil {
+			return c.Err(fmt.Sprintf("Expecting a numeric status code, got '%s'", statusCodeString))
```
Comment (abiosoft) [inline caddyhttp/status/setup.go]: why not just `c.Errf` ?

File: caddyhttp/status/status.go [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+// Status is a middleware to return status code for request
+type Status struct {
+	Rules map[string]int
+	Next  httpserver.Handler
+}
+
+// ServeHTTP implements the httpserver.Handler interface
+func (status Status) ServeHTTP(w http.ResponseWriter, r *http.Request) (int, error) {
+	if statusCode, ok := status.Rules[r.URL.Path]; ok {
+		return statusCode, nil
```
Comment (mholt) [inline caddyhttp/status/status.go]: Hmm, this middleware is an odd exception in the way I've described how the [contract](https://godoc.org/github.com/mholt/caddy/caddyhttp/httpserver#Handler) for a Handler works:

> If ServeHTTP writes to the response body, it should return a status code of 0. This signals to other handlers above it that the response body is already written, and that they should not write to it also.

In this case we don't write a response body but we do essentially "finish" handling the request (which is what I meant by "writes to the response body" - perhaps I should clarify that in the godoc by changing it to "writes the response header").

So what happens here is two-fold:

- If we return statusCode >= 400, great; an error page can be served by the errors middleware.
- If we return statusCode < 400... not so great, because there's no code to write a response body in that case, or even write the status code to the header of the response.

In other words, we need a condition, something like:

```go
if statusCode < 400 {
    w.WriteHeader(statusCode)
    return 0, nil
}
return statusCode, nil
```

**Comment by mholt** [comment]
@evvvvr The latest changes are looking great! Just one more inline comment that I just made and I think it's getting really close. :)

**Comment by evvvvr** [comment]
Changed directive syntax to `status code path` and
```
status code {
   path
   path...
} 
``` 
Is it ok?
Please, be advised that code in PR supports **exact** path only. I propose
```
status code {
  basepath base_path_value
}
```
syntax for it.

**Comment by mholt** [comment]
I just noticed the exact path matching. I think we should choose either base or exact, but probably not both (confusing). I'm leaning towards base path matching since it will allow you to hide a whole folder, and if you just need to hide a file, you can list the full file path.

**Comment by evvvvr** [comment]
Replaced exact path matching with base path check and used [`httpserver.HandlerConfig`](https://godoc.org/github.com/mholt/caddy/caddyhttp/httpserver#HandlerConfig) for middleware configuration.

File: caddyhttp/status/status.go [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+		RequestMatcher: httpserver.PathMatcher(basePath),
+	}
+}
+
+// BasePath implements httpserver.HandlerConfig interface
+func (rule *Rule) BasePath() string {
+	return rule.Base
+}
+
+// Match implements httpserver.HandlerConfig interface
```
Comment (abiosoft) [inline caddyhttp/status/status.go]: This method is redundant since `Rule` embeds `httpserver.RequestMatcher`.

**Comment by abiosoft** [comment]
Finally, looks like this branch is behind master and a rebase is needed.

**Comment by abiosoft** [comment]
Thanks :)

**Comment by mholt** [comment]
Thank you @evvvvr! I'll write up some docs for this.
